libgloss/arc/hl/hl_api.c

```diff
@@ -0,0 +1,319 @@
+    return NULL;
+
+  *type = PAT_INT;
+  *size = 4;
+  *val = x;
+
+  return p + buf_used;
+}
+
+/*
+ * Pack data (pointer and legth) to provided buffer.
+ * Packing structure:
+ *     uint16 type  (PAT_STRING = 4)
+ *     uint16 size  (length)
+ *     char   buf[length]
+ */
+volatile __uncached char *
+hl_pack_ptr (volatile __uncached char *p, const void *s, uint16_t len)
+{
+  volatile __uncached uint16_t *type = (volatile __uncached uint16_t *) p;
+  volatile __uncached uint16_t *size = (volatile __uncached uint16_t *)
+                                       (p + 2);
+  volatile __uncached char *buf = p + 4;
+  const uint32_t buf_used = 4 + ALIGN(len, 4);
+
+  if (hl_buf_left (p) < buf_used)
+    return NULL;
+
+  *type = PAT_STRING;
+  *size = len;
+
+  /* _vdmemcpy(buf, s, len); */
+  for (uint16_t i = 0; i < len; i++)
+    buf[i] = ((const char *) s)[i];
+
+  return p + buf_used;
+}
+
+/*
+ * Pack NUL-terminated string to provided buffer.
+ * Packing structure:
+ *     uint16 type  (PAT_STRING = 4)
+ *     uint16 size  (length)
+ *     char   buf[length]
+ */
+volatile __uncached char *
+hl_pack_str (volatile __uncached char *p, const char *s)
+{
+  return hl_pack_ptr (p, s, strlen (s) + 1);
+}
+
+/* Unpack integer value (uint32_t) from a buffer. */
+volatile __uncached char *
+hl_unpack_int (volatile __uncached char *p, uint32_t *x)
+{
+  volatile __uncached uint16_t *type = (volatile __uncached uint16_t *) p;
+  volatile __uncached uint16_t *size = (volatile __uncached uint16_t *)
+                (p + 2);
+  volatile __uncached uint32_t *val = (volatile __uncached uint32_t *)
+                (p + 4);
+  const uint32_t buf_used = 8;
+
+  if (*type != PAT_INT || *size != 4 || hl_buf_left (p) < buf_used)
+    return NULL;
+
+  *x = *val;
+
+  return p + buf_used;
+}
+
+/* Unpack data from a buffer. */
+volatile __uncached char *
+hl_unpack_ptr (volatile __uncached char *p, void *s, uint32_t *plen)
+{
+  volatile __uncached uint16_t *type = (volatile __uncached uint16_t *) p;
+  volatile __uncached uint16_t *size = (volatile __uncached uint16_t *)
+                                       (p + 2);
+  volatile __uncached char *buf = p + 4;
+  uint32_t buf_used;
+  uint32_t len;
+
+  if (*type != PAT_STRING || hl_buf_left (p) < 4)
+    return NULL;
+
+  len = *size;
+  buf_used = 4 + ALIGN (len, 4);
+
+  if (hl_buf_left (p) < buf_used)
+    return NULL;
+
+  *plen = len;
+
+  /* _vsmemcpy(s, buf, len); */
+  for (uint32_t i = 0; i < len; i++)
+    ((char *) s)[i] = buf[i];
+
+  return p + buf_used;
+}
+
+/*
+ * Unpack data from a buffer.
+ *
+ * No difference compared to hl_unpack_ptr, except that this function
+ * does not return a length.
+ */
+volatile __uncached char *
+hl_unpack_str (volatile __uncached char *p, char *s)
+{
+  uint32_t len;
+  return hl_unpack_ptr (p, s, &len);
+}
+
+/* Return length of packed data (PAT_STRING) if it is on top of the buffer. */
+uint32_t
+hl_get_ptr_len (volatile __uncached char *p)
```

Comment:
please move this function after all the "static" ones

Reply:
done